### PR TITLE
Make TransfiniteInterpolationManifold more robust

### PR DESCRIPTION
**Pull request by kronbichler**
The point identification of the `pull_back` function to the atlas of TransfiniteInterpolationManifold would  sometimes fail for strongly curved, long and skinny cells. This PR fixes this issue, as well as some outside-cell issues.

@tjhei for your information.

### CODE REVIEW

**Comment by tjhei** [comment]
this would be great to have in 9.1.  #8080

File: source/grid/manifold_lib.cc [hunk 1]
```diff
@@ -2026,7 +2027,7 @@ TransfiniteInterpolationManifold<dim, spacedim>::pull_back(
           return chart_point + update;
         }
 
-      // every 8 iterations, including the first time around, we create an
+      // every 9 iterations, including the first time around, we create an
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: I hope the success does not depend on whether we do 8 or 9 iterations. :-)

File: source/grid/manifold_lib.cc [hunk 2]
```diff
@@ -2083,8 +2085,15 @@ TransfiniteInterpolationManifold<dim, spacedim>::pull_back(
...
-      if (alpha < 1e-7)
+      // If alpha got very small, it is likely due to a bad Jacobian
+      // approximation with Broyden's method (relatively far away from the
+      // zero), which can be corrected by the outer loop when a Newton update
+      // is recomputed. The second case is when the Jacobian is actually bad
+      // and we should fail as early as possible. Since we cannot really
+      // distinguish the two, we must continue here in any case.
+      if (alpha <= 1e-4)
         return outside;
+      // must_recompute_jacobian = true;
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: remove?

File: source/grid/manifold_lib.cc [hunk 3]
```diff
@@ -2100,15 +2109,20 @@ TransfiniteInterpolationManifold<dim, spacedim>::pull_back(
...
-      Tensor<1, spacedim> jac_update;
-      for (unsigned int d = 0; d < spacedim; ++d)
-        for (unsigned int e = 0; e < dim; ++e)
-          jac_update[d] += delta_x[e] * inv_grad[d][e];
-      for (unsigned int d = 0; d < spacedim; ++d)
-        for (unsigned int e = 0; e < dim; ++e)
-          inv_grad[d][e] += factor[e] * jac_update[d];
+
+      // prevent division by zero
+      if (delta_x * Jinv_deltaf != 0)
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: is this enough without an eps? You don't run into degenerate numbers?

File: source/grid/manifold_lib.cc [hunk 4]
```diff
@@ -2100,15 +2109,20 @@ TransfiniteInterpolationManifold<dim, spacedim>::pull_back(
...
+        {
+          const Tensor<1, dim> factor =
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: unrelated to this PR, but would you mind adding a comment with the formula here?

File: source/grid/manifold_lib.cc [hunk 5]
```diff
@@ -2329,6 +2343,68 @@ TransfiniteInterpolationManifold<dim, spacedim>::compute_chart_points(
            (use_structdim_2_guesses ^ use_structdim_3_guesses),
          ExcInternalError());
 
+
+
+  auto compute_chart_point = [&](const typename Triangulation<dim, spacedim>::
+                                   cell_iterator &  cell,
+                                 const unsigned int i) {
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: i is the point index?

File: source/grid/manifold_lib.cc [hunk 6]
```diff
@@ -2329,6 +2343,68 @@ TransfiniteInterpolationManifold<dim, spacedim>::compute_chart_points(
...
+      guess = chart_points[1] + (chart_points[2] - chart_points[0]);
+    else if (use_structdim_2_guesses && 3 < i)
+      guess = guess_chart_point_structdim_2(i);
+    else if (use_structdim_3_guesses && 4 < i)
+      guess = guess_chart_point_structdim_3(i);
+    else if (dim == 3 && i > 7 && surrounding_points.size() == 26)
+      {
+        if (i < 20)
+          guess =
+            0.5 *
```
Comment (tjhei) [inline source/grid/manifold_lib.cc]: do 0.5*(a+b)?

**Comment by tjhei** [comment]
@kronbichler our three cylinder examples still fails (but now in assembly):
```
An error occurred in line <1948> of file </scratch/deal-git/source/grid/manifold_lib.cc> in function
    dealii::Point<spacedim> dealii::TransfiniteInterpolationManifold<dim, spacedim>::push_forward(const typename dealii::Triangulation<dim, spacedim>::cell_iterator&, const dealii::Point<dim>&) const [with int dim = 3; int spacedim = 3; typename dealii::Triangulation<dim, spacedim>::cell_iterator = dealii::TriaIterator<dealii::CellAccessor<3, 3> >]
The violated condition was:
    GeometryInfo<dim>::is_inside_unit_cell(chart_point, 1e-6)
Additional information:
    chart_point is not in unit interval
```

File: source/grid/manifold_lib.cc (hunk 1 above)
Comment (kronbichler) [reply source/grid/manifold_lib.cc]: No, but I observed it was faster on average... :smile:

File: source/grid/manifold_lib.cc (hunk 3 above)
Comment (kronbichler) [reply source/grid/manifold_lib.cc]: It is unlikely, but let us play safe.

**Comment by kronbichler** [comment]
Let me check the three cylinders again.

**Comment by kronbichler** [comment]
> our three cylinder examples still fails (but now in assembly)

It turns out I had forgotten to adapt the tolerance in a second place. Now it should work. The problem with the three cylinders is that the way define manifolds there is a bit delicate which causes the transfinite interpolation to be tricky. The problem is that we have the cylindrical manifolds joining, and I think the cylindrical descriptions as we use them generate some overlap. I tried to fix those with some ellipse description but that is not fully stable either. In any case, I have extracted two scenarios from the cylinders into `transfinite_manifold_{08,09}` test cases which cover some of the difficulties. Not all, but I was a bit reluctant to include the real test case which is very expensive and has a large mesh. If you want, we can include it.

Regarding the release: While I believe the current status is the right thing to do, it involves non-trivial code changes and I would be careful with that.

**Comment by tjhei** [comment]
awesome!

> Not all, but I was a bit reluctant to include the real test case which is very expensive and has a large mesh. If you want, we can include it.

I think we are okay without this test.

> Regarding the release: While I believe the current status is the right thing to do, it involves non-trivial code changes and I would be careful with that.

Depends on how close we are to making the release. I would love to have this in there, but I understand your hesitation. My only worry is that it might introduce warnings/errors with some compilers.

**Comment by kronbichler** [comment]
Since it's my patch, I'll let other take the decision whether this is safe enough now, especially @tamiko. I am happy to re-run it on my machine and fix any follow-up issue tomorrow, to the extent possible.

**Comment by tamiko** [comment]
@tjhei @kronbichler I wouldn't tag a release before Tuesday next week. Better a bit too much testing time than too little. That said, we could take over #8191 into rc2 (which I want to tag this Evening and see how the testsuite responds. We can always revert in rc3.

**Comment by kronbichler** [comment]
That's a good plan. In any case, this would imply moving the changelog around. I won't get to do it today any more, though.

**Comment by tamiko** [comment]
the following tests fail for at least one configuration:

 - [ ] (a) adolc/helper_scalar_failure_04.release
 - [ ] (a) adolc/helper_tape_index_01.release
 - [ ] (a) algorithms/general_data_storage_01.debug
 - [ ] (a) algorithms/general_data_storage_01.release
 - [x] <s>(1) fe/fe_enriched_color_07.mpirun=2.debug</s> (likely clang-4 bug)
 - [x] <s>(1) fe/fe_enriched_color_07.mpirun=2.release</s> (likely clang-4 bug)
 - [ ] (1) particles/particle_handler_08.debug
 - [ ] (1) particles/particle_handler_08.release
 - [ ] (a) rol/vector_adaptor_02.debug
 - [ ] (a) rol/vector_adaptor_02.release

We should either make them more robust (

**Comment by masterleinad** [comment]
@tamiko Did you mean to put the table here?

**Pull request merged**